src/back/hlsl/writer.rs

```diff
@@ -1663,25 +1663,35 @@ impl<'a, W: fmt::Write> super::Writer<'a, W> {
                         }
                     }
 
+                    // FXC doesn't support fallthrough so we duplicate the body of the following case blocks
                     if case.fall_through {
-                        // Generate each fallthrough case statement in a new block. This is done to
-                        // prevent symbol collision of variables declared in these cases statements.
-                        writeln!(self.out, "{}/* fallthrough */", indent_level_2)?;
-                        writeln!(self.out, "{}{{", indent_level_2)?;
-                    }
-                    for sta in case.body.iter() {
-                        self.write_stmt(
-                            module,
-                            sta,
-                            func_ctx,
-                            back::Level(indent_level_2.0 + usize::from(case.fall_through)),
-                        )?;
-                    }
+                        let curr_len = i + 1;
+                        let end_case_idx = curr_len
+                            + cases
+                                .iter()
+                                .skip(curr_len)
+                                .position(|case| !case.fall_through)
+                                .unwrap();
+                        let indent_level_3 = indent_level_2.next();
+                        for case in &cases[i..=end_case_idx] {
+                            writeln!(self.out, "{}{{", indent_level_2)?;
+                            for sta in case.body.iter() {
+                                self.write_stmt(module, sta, func_ctx, indent_level_3)?;
+                            }
+                            writeln!(self.out, "{}}}", indent_level_2)?;
+                        }
```

Comment:
I find the naming to be slightly more confusing here since `rest` and `fallthrough` contain the first case as well. Getting `fallthrough_count` should also skip the first element (since it's the first case) and add one at the end.

I tried to make it more clear as well without much luck...